libindy/src/commands/mod.rs

```diff
@@ -112,77 +117,120 @@ impl CommandExecutor {
                         }
-                        Ok(Command::NonSecrets(cmd)) => {
+                        Some(Command::NonSecrets(cmd)) => {
                             debug!("NonSecretCommand command received");
                             non_secret_command_executor.execute(cmd);
                         }
-                        Ok(Command::Payments(cmd)) => {
+                        Some(Command::Payments(cmd)) => {
                             debug!("PaymentsCommand command received");
-                            payments_command_executor.execute(cmd);
+                            payments_command_executor.execute(cmd).await;
                         }
-                        Ok(Command::Cache(cmd)) => {
+                        Some(Command::Cache(cmd)) => {
                             debug!("CacheCommand command received");
-                            cache_command_executor.execute(cmd);
+                            cache_command_executor.execute(cmd).await;
                         }
-                        Ok(Command::Exit) => {
+                        Some(Command::Exit) => {
                             debug!("Exit command received");
-                            break
+                            return true
                         }
-                        Err(err) => {
-                            error!("Failed to get command!");
-                            panic!("Failed to get command! {:?}", err)
+                        None => {
+                            warn!("No command to execute");
+                        }
+                    }
+
+                    false
+                };
+
+                let mut in_progress_tasks = FuturesUnordered::new();
+                loop {
+                    trace!("CommandExecutor main loop >>");
+                    let mut break_main_loop = false;
+                    block_on(async {
+                        trace!("CommandExecutor async block");
+                        select! {
+                            cmd = receiver.next() => {
+                                trace!("CommandExecutor::select new command");
+                                let in_progress_task = _exec_cmd(cmd, &anoncreds_command_executor, &crypto_command_executor, &ledger_command_executor, &pool_command_executor, &did_command_executor, &wallet_command_executor, &pairwise_command_executor, &blob_storage_command_executor, &non_secret_command_executor, &payments_command_executor, &cache_command_executor);
+                                in_progress_tasks.push(in_progress_task);
+                            }
+                            should_complete_main_loop = in_progress_tasks.next() => {
+                                trace!("CommandExecutor::select in progress task, break loop {:?}", should_complete_main_loop);
+                                break_main_loop = should_complete_main_loop.unwrap_or(false);
+                            }
+                            complete => {
```

Comment:
what case this does event happen?

Reply:
in case of both streams are finished. Probably the select may be replaced by smth else and the main loop can be replaced to some while. I keep current version as the most easy to read from my list, but I'm open to any proposal how to improve it more